### PR TITLE
Add `FUNCTION LIST` command.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
https://redis.io/docs/latest/commands/function-list/

Depends/based on #270 

TODOs:
- [x] response conversion tests
- [x] clean-up and/or update response conversion
- [x] UT
- [x] IT

### CODE REVIEW

File: java/integTest/src/test/java/glide/cluster/ClusterTransactionTests.java [hunk 1]
```diff
@@ -104,7 +105,7 @@ public void keyless_transactions_with_group_of_commands(
 
         SingleNodeRoute route = new SlotIdRoute(1, SlotType.PRIMARY);
         Object[] results = clusterClient.exec(transaction, route).get();
-        assertArrayEquals(expectedResult, results);
+        assertDeepEquals(expectedResult, results, true);
```
Comment (tjzhang-BQ) [inline java/integTest/src/test/java/glide/cluster/ClusterTransactionTests.java:108]: should the assertion on line 95 be replaced too?

File: java/integTest/src/test/java/glide/standalone/TransactionTests.java [hunk 2]
```diff
@@ -116,7 +117,7 @@ public void keyless_transactions_with_group_of_commands(
         Object[] expectedResult = builder.apply(transaction);
 
         Object[] results = client.exec(transaction).get();
-        assertArrayEquals(expectedResult, results);
+        assertDeepEquals(expectedResult, results, true);
```
Comment (tjzhang-BQ) [inline java/integTest/src/test/java/glide/standalone/TransactionTests.java:120]: same question as above with the assertion method in `transactions_with_group_of_commands`

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java [hunk 3]
```diff
@@ -78,4 +79,216 @@ public interface ScriptingAndFunctionsClusterCommands {
      * }</pre>
      */
     CompletableFuture<String> functionLoadWithReplace(String libraryCode, Route route);
+
+    /**
+     * Returns information about the functions and libraries.<br>
+     * The command will be routed to all primary nodes.
+     *
+     * @since Redis 7.0 and above
+     * @see <a href="https://redis.io/docs/latest/commands/function-list/">redis.io</a> for details.
```
Comment (tjzhang-BQ) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:88]: question: should we be linking redis or valkey right now? or does it not matter, i remember some change of replacing all redis to valkey that was talked about

File: java/client/src/main/java/glide/api/models/commands/function/FunctionListOptions.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+/** Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0 */
+package glide.api.models.commands.function;
+
+/**
+ * Option for <code>FUNCTION LIST</code> command.
```
Comment (tjzhang-BQ) [inline java/client/src/main/java/glide/api/models/commands/function/FunctionListOptions.java:5]: should it be {@link} for the command name here?

File: java/integTest/src/test/java/glide/TestUtilities.java [hunk 5]
```diff
@@ -76,4 +81,97 @@ public static Map<String, String> parseInfoResponseToMap(String serverInfo) {
...
+            var expectedMap = (Map<?, ?>) expected;
+            var actualMap = (Map<?, ?>) actual;
+            assertEquals(expectedMap.size(), actualMap.size());
+            assertDeepEquals(expectedMap.keySet(), actualMap.keySet(), false);
+            for (var key : expectedMap.keySet()) {
+                assertDeepEquals(expectedMap.get(key), actualMap.get(key), orderMatters);
+            }
+        } else {
+            assertEquals(expected, actual);
+        }
```
Comment (tjzhang-BQ) [inline java/integTest/src/test/java/glide/TestUtilities.java:136]: reminder to double-check the test results with the changes in the transaction IT pr to be safe?

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 6]
```diff
@@ -702,16 +732,37 @@ public void functionLoad(boolean withRoute) {
...
+                (withRoute
+                                ? clusterClient.functionListWithCode(libName, route)
+                                : clusterClient.functionListWithCode(libName))
+                        .get();
+        checkFunctionListResponse(
+                flist, libName, expectedDescription, expectedFlags, Optional.of(newCode));
+
         // TODO test with FCALL
+
+        // TODO FUNCTION FLUSH at the end
```
Comment (tjzhang-BQ) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:766]: are these just reserving space for commands to be done later?

**Review by tjzhang-BQ — APPROVED** [review]

File: java/integTest/src/test/java/glide/cluster/CommandTests.java (hunk 6 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/cluster/CommandTests.java:766]: Yes, I removed this TODO in the next PR

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:88]: I don't care about links - will update all of them later in bulk.

File: glide-core/src/client/value_conversion.rs [hunk 7]
```diff
@@ -287,6 +290,110 @@ pub(crate) fn convert_to_expected_type(
...
+           ...
+        */
+        ExpectedReturnType::ArrayOfMapsRecursive => match value {
+            // empty array, or it is already contains a map (RESP3 response) - no conversion needed
+            Value::Array(ref array) if array.is_empty() || matches!(array[0], Value::Map(_)) => {
+                Ok(value)
+            }
+            Value::Array(array) => {
+                convert_array_of_flat_maps(array, Some(ExpectedReturnType::ArrayOfMaps))
+                /*
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply glide-core/src/client/value_conversion.rs:331]: Yes, TBD which implementation to keep.

File: java/client/src/main/java/glide/api/models/commands/function/FunctionListOptions.java (hunk 4 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/commands/function/FunctionListOptions.java]: Added

**Pull request merged**